closed/adds/jdk/src/share/classes/sun/security/ec/NativeECDHKeyAgreement.java

```diff
@@ -0,0 +1,168 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * ===========================================================================
+ * (c) Copyright IBM Corp. 2022, 2022 All Rights Reserved
+ * ===========================================================================
+ */
+
+package sun.security.ec;
+
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.ProviderException;
+import java.security.SecureRandom;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.ECParameterSpec;
+
+import javax.crypto.KeyAgreementSpi;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+
+import jdk.crypto.jniprovider.NativeCrypto;
+
+public final class NativeECDHKeyAgreement extends KeyAgreementSpi {
+
+    /* private key, if initialized */
+    private ECPrivateKeyImpl privateKey;
+
+    /* public key, non-null between doPhase() & generateSecret() only */
+    private ECPublicKeyImpl publicKey;
+
+    /* length of the secret to be derived */
+    private int secretLen;
+
+    private static final NativeCrypto nativeCrypto = NativeCrypto.getNativeCrypto();
+
+    /**
+     * Constructs a new NativeECDHKeyAgreement.
+     */
+    public NativeECDHKeyAgreement() {
+    }
+
+    @Override
+    protected void engineInit(Key key, SecureRandom random)
+            throws InvalidKeyException {
+        if (!(key instanceof PrivateKey)) {
+            throw new InvalidKeyException
+                        ("Key must be instance of PrivateKey");
+        }
+        this.privateKey = (ECPrivateKeyImpl) ECKeyFactory.toECKey(key);
+        this.publicKey = null;
+    }
+
+    @Override
+    protected void engineInit(Key key, AlgorithmParameterSpec params, SecureRandom random)
+            throws InvalidKeyException, InvalidAlgorithmParameterException {
+        if (params != null) {
+            throw new InvalidAlgorithmParameterException
+                        ("Parameters not supported");
+        }
+        engineInit(key, random);
+    }
+
+    @Override
+    protected Key engineDoPhase(Key key, boolean lastPhase)
+            throws InvalidKeyException, IllegalStateException {
+        if (this.privateKey == null) {
+            throw new IllegalStateException("Not initialized");
+        }
+        if (this.publicKey != null) {
+            throw new IllegalStateException("Phase already executed");
+        }
+        if (!lastPhase) {
+            throw new IllegalStateException
+                ("Only two party agreement supported, lastPhase must be true");
+        }
+        if (!(key instanceof ECPublicKeyImpl)) {
+            throw new InvalidKeyException
+                ("Key must be a ECPublicKeyImpl");
```

Comment:
I think the old message was better: `"Key must be a PublicKey with algorithm EC"`.

Reply:
For symmetry, perhaps this should also use `ECKeyFactory.toECKey()` like the private key code above.

Reply:
done